### ISSUE TITLE
Adding fabric components CSS breaks web part drop down properties

### ISSUE THREAD

**Issue by shaipetel** [SharePoint/sp-dev-docs#434]
#### Category
- [ ] Question
- [ ] Typo
- [x] Bug
- [ ] Enhancement / Suggestion
- [ ] Additional article idea

#### Expected or Desired Behavior
Adding a reference to fabric component CSS to be able to use ms-list classes.
Using components like list or pivot, on a web part that has a PropertyPaneDropdown property should still be able to work and user should be able to use the dropdown.

#### Observed Behavior
Once adding a reference to the components CSS, when opening the web part properties pane the user can no longer use the drop down properties. The dropdown control stopped working.

#### Steps to Reproduce
Using RC0, running on the local workbench.
Create a web part that has a simple dropdown property using PropertyPaneDropdown.
Make sure the dropdown works when editing the web part.
Add a reference to fabric components CSS:
Option 1 I tried:
SPComponentLoader.loadCss("https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.min.css");
SPComponentLoader.loadCss("https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.components.min.css");

Option 2 I tried:
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.min.css');
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.components.min.css');

Option 3 I tried:
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/1.0/fabric.min.css');
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/1.0/fabric.components.min.css');

No matter which option you try, one you add a reference to this CSS you can no longer us the drop down in the web part properties.
Try to edit web part properties, you cannot open the drop down.

**Comment by manishgarg1** [comment]
Hi @shaipetel, fabric.component.css is deprecated. Developers are not expected to use it. Secondly, why are you having to laod the fabric.css manually? Its already on the page.

**Comment by shaipetel** [comment]
Hi @manishgarg1 
Thanks for replying... been 12 days since I posted it and no one bothered replying.

Can you please explain since when is components deprecated?
If I simply go to office fabric home page: https://dev.office.com/fabric
And click for Fabric JS: https://github.com/OfficeDev/office-ui-fabric-js
It says right there to use:
<link rel="stylesheet" href="https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.min.css">
<link rel="stylesheet" href="https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.components.min.css">

Where does it say it is deprecated?

Even if I follow the link to the latest 2.6.1 link:
https://github.com/OfficeDev/office-ui-fabric-core/blob/master/ghdocs/V2GETSTARTED.md

It still say to register a CSS to components! I tried these one as well:
<link rel="stylesheet" href="https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.min.css">
<link rel="stylesheet" href="https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.components.min.css">

Assuming you are right and they are deprecated, if I want to use office fabric components on SPFx using KnockoutJS or clean JS, **not angular and not react**, what other option do I have?

Please keep this classified as a bug, at least until there is a clear answer to this, since currently office fabric does not state these are deprecated anywhere, and there is no clear newer components to use.

Thanks,

**Comment by manishgarg1** [comment]
HI @shaipetel, by deprecated I mean from SP pages which are React pages. You do have a valid point. I will circle back with the Fabric team and get back to you.

**Comment by manishgarg1** [comment]
Hi @shaipetel, one more thing. We are switching all Fabric React components to use css modules. That will by design resolve this issue.

**Comment by shaipetel** [comment]
Well, I can't use react since I have some functional requirements for my web part (like dynamic rendering templates). Angular is an overkill and will complicate my template rendering and binding. React is incapable of delivering dynamic templates at all.

If there is a different version of the CSS I can use - let me know. This is now stopping our release due to broken property panel. Otherwise I will wait for a solution.

Thanks!

**Comment by manishgarg1** [comment]
Hi @shaipetel, What are you using for dynamic templates? handlebars? And yes, Angular is an overkill for that. I will talk with the design team today. Did you look and see what styles in component.css break the drop down? For the time being, to unblock your release, could you override them in your app? By out next drop, this should get fixed by design and you can undo your overrides.

**Comment by shaipetel** [comment]
I use knockout, it is perfect for my web part. I'm not entirely sure what in the CSS causes this but its not just the dropdowns that are broken - all buttons text are also not aligned correctly.

**Comment by shaipetel** [comment]
I noticed I can click on the drop down, and even that it doesn't open I can use the keyboard up/down arrows to change selection. So it appears the drop down is not completely destroyed but more likely just hidden somehow by wrong styles.

**Comment by srideshpande** [comment]
@shaipetel - Could you please share a screenshot of how your property pane looks like? and also how does your package.json looks?

It would be great if you could also share the code you are using to communicate to property pane from your web part.

**Comment by shaipetel** [comment]
Sure, code sample:
`  public constructor() {
    super();
    //option 1
SPComponentLoader.loadCss("https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.min.css");
SPComponentLoader.loadCss("https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/css/fabric.components.min.css");
SPComponentLoader.loadScript("https://static2.sharepointonline.com/files/fabric/office-ui-fabric-js/1.2.0/js/fabric.min.js");
    //option 2
SPComponentLoader.loadCss("https://appsforoffice.microsoft.com/fabric/1.0/fabric.min.css");
SPComponentLoader.loadCss("https://appsforoffice.microsoft.com/fabric/1.0/fabric.components.min.css");
    ///option 3
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.min.css');
SPComponentLoader.loadCss('https://appsforoffice.microsoft.com/fabric/2.6.1/fabric.components.min.css');
  }`

Leave only 1 option active at a time, the problem will happen with any CSS pair you load. In fact, it will happen if you load any of the fabric.components.min.css files, you don't have to load the fabric.min.css

For the drop down, I have a simple drop down property:
`        wp.PropertyPaneDropdown('dataSource.type', {
          label: "Data source type:",
          disabled: self.propsHelper.dataSource.typeSelected,
          options: self.propsHelper.dataSourceTypesOptions
          //, selectedKey: self.propsHelper.dataSourceTypes.SPList
        })`
(Notice I don't use selectedKey? since it also doesn't work properly, but that's another issue)

The result will be: the drop down will appear, but you won't be able to open it.
I attached how it looks like when you click it - there is a thin shadow under it and nothing happens.
![dropdown opened](https://cloud.githubusercontent.com/assets/6225881/23561759/9fd835de-000d-11e7-8d52-72919baa17b6.png)

However, you can then use the up/down arrows in the keyboard to select different values.

Hope this helps!

**Comment by srideshpande** [comment]
@shaipetel  - Thanks for the info. Tested it myself and seeing the issue. The `components.css` file is the culprit here. Styles in that file are out of date and hence the issue. We are not recommending to use that file anymore, we will update the instructions page soon.

Will update you with new instructions when they become ready.

Thanks for bringing it to our attention.

**Comment by patmill** [comment]
Doc around working with fabric now and in the near future is here - https://dev.office.com/sharepoint/docs/spfx/web-parts/guidance/office-ui-fabric-integration.

**Comment by shaipetel** [comment]
Thanks @patmill I see there are no good news for me there yet, anyway I came up with my own solution ([read about it in my blog](https://kwizcom.blogspot.ca/2017/03/using-office-ui-fabric-js-in-spfx.html))

Basically, I rename the classes and JS object with my own prefix.

I built a service every developer can use to generage the office ui fabric JS with components here:
https://apps.kwizcom.com/libs/office-ui-fabric-js/1.4.0/css/fabric.css?prefix=kw
https://apps.kwizcom.com/libs/office-ui-fabric-js/1.4.0/css/fabric.components.css?prefix=kw
https://apps.kwizcom.com/libs/office-ui-fabric-js/1.4.0/js/fabric.js?prefix=kw

Anyone in the same situation is welcome to use our CDN, and use his own prefix.

I can't think of a better way currently to handle this. I guess if everyone just generate their own version with their own prefix it would pretty much solve this issue.

I'll keep an eye on that link you shared to see if they post anything there in the future.

**Comment by manishgarg1** [comment]
Hi @shaipetel, yes for fabric.css and fabric.component.css, using your own name-prefixes is the only way that a developer can prevent the web part from breaking frequently or the web part breaking the page styles around it.

**Comment by manishgarg1** [comment]
Hi @shaipetel, looks like you are unblocked for now. Hence I am closing the issue. The final update for fabric react integration and web parts will be published soon. 

Thanx for your feedback and support. If you still see the issue in production, you can always reopen this issue or submit a new one.

**Comment by praveenbattula** [comment]
Even today I see the issue with 2.6.1 fabric components. Is there any update at all on this?
